connector/src/main/java/org.apache.flink/connector/nebula/catalog/NebulaCatalog.java

```diff
@@ -0,0 +1,240 @@
+ */
+
+package org.apache.flink.connector.nebula.catalog;
+
+import static org.apache.flink.connector.nebula.table.NebulaDynamicTableFactory.GRAPH_SPACE;
+import static org.apache.flink.connector.nebula.table.NebulaDynamicTableFactory.IDENTIFIER;
+import static org.apache.flink.connector.nebula.table.NebulaDynamicTableFactory.LABEL_NAME;
+import static org.apache.flink.connector.nebula.table.NebulaDynamicTableFactory.METAADDRESS;
+import static org.apache.flink.connector.nebula.table.NebulaDynamicTableFactory.PASSWORD;
+import static org.apache.flink.connector.nebula.table.NebulaDynamicTableFactory.USERNAME;
+import static org.apache.flink.table.factories.FactoryUtil.CONNECTOR;
+
+import com.facebook.thrift.TException;
+import com.vesoft.nebula.client.graph.data.HostAddress;
+import com.vesoft.nebula.client.meta.MetaClient;
+import com.vesoft.nebula.client.meta.exception.ExecuteFailedException;
+import com.vesoft.nebula.meta.ColumnDef;
+import com.vesoft.nebula.meta.EdgeItem;
+import com.vesoft.nebula.meta.IdName;
+import com.vesoft.nebula.meta.PropertyType;
+import com.vesoft.nebula.meta.Schema;
+import com.vesoft.nebula.meta.TagItem;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.commons.collections.map.HashedMap;
+import org.apache.flink.connector.nebula.utils.DataTypeEnum;
+import org.apache.flink.connector.nebula.utils.NebulaConstant;
+import org.apache.flink.connector.nebula.utils.NebulaUtils;
+import org.apache.flink.table.api.DataTypes;
+import org.apache.flink.table.api.TableSchema;
+import org.apache.flink.table.catalog.CatalogBaseTable;
+import org.apache.flink.table.catalog.CatalogDatabase;
+import org.apache.flink.table.catalog.CatalogDatabaseImpl;
+import org.apache.flink.table.catalog.CatalogTableImpl;
+import org.apache.flink.table.catalog.ObjectPath;
+import org.apache.flink.table.catalog.exceptions.CatalogException;
+import org.apache.flink.table.catalog.exceptions.DatabaseNotExistException;
+import org.apache.flink.table.catalog.exceptions.TableNotExistException;
+import org.apache.flink.table.types.DataType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class NebulaCatalog extends AbstractNebulaCatalog {
+
+    private static final Logger LOG = LoggerFactory.getLogger(NebulaCatalog.class);
+    private final List<HostAddress> hostAndPorts;
+    private final MetaClient metaClient;
+
+    public NebulaCatalog(String catalogName, String defaultDatabase, String username, String pwd,
+                         String address) {
+        super(catalogName, defaultDatabase, username, pwd, address);
+        this.hostAndPorts = NebulaUtils.getHostAndPorts(address);
+        this.metaClient = new MetaClient(hostAndPorts);
+    }
+
+    @Override
+    public List<String> listDatabases() throws CatalogException {
+        Map<String, Integer> spaceMap = new HashMap<>();
```

Comment:
its better to set the init size of hashmap here

Reply:
Yes，it's better to set map's size if we know the number of elements, but spaces' number is unknown here.